core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/ChecksumCalculatingAsyncRequestBody.java

```diff
@@ -225,4 +238,48 @@ public void onComplete() {
             wrapped.onComplete();
         }
     }
+
+    private Iterable<ByteBuffer> bufferAndCreateChunks(ByteBuffer buffer) {
+        int startPosition = 0;
+        int currentBytesRead = buffer.remaining();
+
+        List<ByteBuffer> resultBufferedList = new ArrayList<>();
```

Comment:
Should we buffer the `ByteBuffer`s from upstream? What happens if this gets called from different threads? For example, upstream onNext gets invoked sequentially by two different threads?

This is how event stream messages are buffered: https://github.com/awslabs/aws-eventstream-java/blob/master/src/main/java/software/amazon/eventstream/MessageDecoder.java#L99

Reply:
Done, thanks for the reference code.